From e32b0213a3c733d4775d472bb5dc87c32487904b Mon Sep 17 00:00:00 2001
From: Kishon Vijay Abraham I <kishon@ti.com>
Date: Tue, 2 Aug 2016 14:37:04 +0530
Subject: [PATCH 141/347] mmc: host: omap_hsmmc: unmap sg list for non block
 adma transfers

This patch comes from:
  git://git.ti.com/processor-sdk/processor-sdk-linux.git

dma_unmap_sg() is not invoked for commands which has a data response
(like CMD51) if adma is used. This results in MMC core not seeing
a valid data response for commands like CMD51 causing enumeration
failure. Fix it here by invoking dma_unmap_sg() after transfer
complete interrupt is received.

For block data transfer commands dma_unmap_sg() is invoked in post_req()
callback (for both adma and system dma). In the case of system dma, for
commands like CMD51, dma_unmap_sg() is invoked in
omap_hsmmc_dma_callback() and hence this issue shows-up only with
master DMA, and not slave DMA.

Signed-off-by: Kishon Vijay Abraham I <kishon@ti.com>
Signed-off-by: Sekhar Nori <nsekhar@ti.com>
(cherry picked from commit 5a9cd13103eb203ca8dca09ca5864642b1ff25be)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/mmc/host/omap_hsmmc.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/drivers/mmc/host/omap_hsmmc.c b/drivers/mmc/host/omap_hsmmc.c
index 5703c1c..d5de085 100644
--- a/drivers/mmc/host/omap_hsmmc.c
+++ b/drivers/mmc/host/omap_hsmmc.c
@@ -1130,6 +1130,10 @@ omap_hsmmc_xfer_done(struct omap_hsmmc_host *host, struct mmc_data *data)
 		return;
 	}
 
+	if (host->use_adma && !data->host_cookie)
+		dma_unmap_sg(host->dev, data->sg, data->sg_len,
+			     omap_hsmmc_get_dma_dir(host, data));
+
 	host->data = NULL;
 
 	if (!data->error)
-- 
1.7.5.4

